### ISSUE TITLE
dotnet-openapi refresh does not truncate json documents.

### ISSUE THREAD

**Issue by chrisbecke** [dotnet/aspnetcore#35767]
### Describe the bug
When dotnet-openapi is used to refresh a swagger.json document, if the new document is shorter than the old document, the file is not truncated to the new size leaving invalid trailing bytes that render it un-parsable, causing build errors.

### To Reproduce

1. Create a new webapi project and add 2 [HttpXXX] methods to the controller.
2. Run `dotnet-openapi add url https://localhost:5001/swagger/v1/swagger.json --output-file MyApi.json` from a terminal.
3. Rebuild to verify the MyApiClient is accessible and works.
4. Remove a method from the controller and rebuild and rerun.
5. Run `dotnet-openapi refresh https://localhost:5001/swagger/v1/swagger.json`
6. Verify that the MyApi.json file retains its original file and has garbage data trailing the new, shorter, swagger definition.

### Further technical details

```bash
user@mac % dotnet tool list --global
Package Id                       Version      Commands                   
-------------------------------------------------------------------------
dotnet-aspnet-codegenerator      2.2.3        dotnet-aspnet-codegenerator
dotnet-ef                        5.0.9        dotnet-ef                  
dotnet-svcutil                   2.0.2        dotnet-svcutil             
microsoft.dotnet-openapi         5.0.9        dotnet-openapi             
```

**Comment by bradygaster** [comment]
cc @darrelmiller do you have an associated issue related to this in Microsoft.OpenAPI? If not, @dougbu do you know if this is a downstream issue unrelated to Microsoft.OpenAPI?

**Comment by darrelmiller** [comment]
I am not aware of such an issue with Microsoft.OpenAPI.NET , although I'd need to see how the tool uses OpenAPI.NET to see if it is something that it is doing wrong or the calling code is doing wrong.

**Comment by bradygaster** [comment]
@dougbu your assistance on this would be helpful as you're our resident expert on dotnet-openapi afaik.

**Comment by dougbu** [comment]
I'm confused because the `dotnet-openapi refresh` command should never download from more than a single URI and shouldn't use either `dotnet-aspnet-codegenerator` or cache anything from previous `dotnet-openapi` executions. But, it's possible there's something wonky with how we open an existing file or close the `FileStream`.

I would appreciate a test project if one already exists. Could you please upload yours to GitHub @chrisbecke❔

**Comment by dougbu** [comment]
> do you know if this is a downstream issue unrelated to Microsoft.OpenAPI?

&hellip;

> I'd need to see how the tool uses OpenAPI.NET to see if it is something that it is doing wrong or the calling code is doing wrong.

The tool isn't using OpenAPI.NET as far as I know. It's just downloading OpenAPI docs from the network (loopback network in this case) and writing them out.

**Comment by dougbu** [comment]
By the way @chrisbecke, does your step 5 include `--output-file MyApi.json` i.e. reuse the same output file as step 2❔

**Comment by chrisbecke** [comment]
This is exactly what I did:

```bash
mkdir dotnet-openid-bug
cd dotnet-openid-bug
dotnet new sln
dotnet new webapi -o webapi
dotnet sln add webapi
open dotnet-openid-bug.sln
```

Visual Studio opened the project. I right clicked the project and added a new class picking ApiController template to get a ValuesController.cs 
I then clicked run, and it opened in the browser as https://localhost:5001/swagger/index.html
While its running

```bash
cd webapi
dotnet-openapi add url https://localhost:5001/swagger/v1/swagger.json --output-file WebApi.json
```

Now we stop the project, and delete the existing WeatherController.cs and related model file from the root as the scaffolded file will conflict. And its convenient to make the next swagger.json smaller.
Rebuild and run.

```bash
dotnet-openapi refresh https://localhost:5001/swagger/v1/swagger.json
```
 Now examine the WebApi.json file.

```json
{
  "openapi": "3.0.1",
  "info": {
    "title": "webapi",
    "version": "v1"
  },
  "paths": {
    "/api/Values": {
      "get": {
        "tags": [
          "Values"
        ],
        "responses": {
          "200": {
            "description": "Success",
            "content": {
              "text/plain": {
                "schema": {
                  "type": "array",
                  "items": {
                    "type": "string"
                  }
                }
              },
              "application/json": {
                "schema": {
                  "type": "array",
                  "items": {
                    "type": "string"
                  }
                }
              },
              "text/json": {
                "schema": {
                  "type": "array",
                  "items": {
                    "type": "string"
                  }
                }
              }
            }
          }
        }
      },
      "post": {
        "tags": [
          "Values"
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "type": "string",
                "nullable": true
              }
            },
            "text/json": {
              "schema": {
                "type": "string",
                "nullable": true
              }
            },
            "application/*+json": {
              "schema": {
                "type": "string",
                "nullable": true
              }
            }
          }
        },
        "responses": {
          "200": {
            "description": "Success"
          }
        }
      }
    },
    "/api/Values/{id}": {
      "get": {
        "tags": [
          "Values"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "integer",
              "format": "int32"
            }
          }
        ],
        "responses": {
          "200": {
            "description": "Success",
            "content": {
              "text/plain": {
                "schema": {
                  "type": "string"
                }
              },
              "application/json": {
                "schema": {
                  "type": "string"
                }
              },
              "text/json": {
                "schema": {
                  "type": "string"
                }
              }
            }
          }
        }
      },
      "put": {
        "tags": [
          "Values"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "integer",
              "format": "int32"
            }
          }
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "type": "string",
                "nullable": true
              }
            },
            "text/json": {
              "schema": {
                "type": "string",
                "nullable": true
              }
            },
            "application/*+json": {
              "schema": {
                "type": "string",
                "nullable": true
              }
            }
          }
        },
        "responses": {
          "200": {
            "description": "Success"
          }
        }
      },
      "delete": {
        "tags": [
          "Values"
        ],
        "parameters": [
          {
            "name": "id",
            "in": "path",
            "required": true,
            "schema": {
              "type": "integer",
              "format": "int32"
            }
          }
        ],
        "responses": {
          "200": {
            "description": "Success"
          }
        }
      }
    }
  },
  "components": { }
}
      "get": {
        "tags": [
          "WeatherForecast"
        ],
        "responses": {
          "200": {
            "description": "Success",
            "content": {
              "text/plain": {
                "schema": {
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/WeatherForecast"
                  }
                }
              },
              "application/json": {
                "schema": {
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/WeatherForecast"
                  }
                }
              },
              "text/json": {
                "schema": {
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/WeatherForecast"
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "WeatherForecast": {
        "type": "object",
        "properties": {
          "date": {
            "type": "string",
            "format": "date-time"
          },
          "temperatureC": {
            "type": "integer",
            "format": "int32"
          },
          "temperatureF": {
            "type": "integer",
            "format": "int32",
            "readOnly": true
          },
          "summary": {
            "type": "string",
            "nullable": true
          }
        },
        "additionalProperties": false
      }
    }
  }
}
```

**Comment by dougbu** [comment]
Thanks @chrisbecke I'll likely not get to this today but appreciate the extra information

**Comment by dougbu** [comment]
> Would uploading the project to github still be helpful?

Not really. Your repro steps are straightforward and less time-consuming than I expected. Just need to clear enough time to debug things.

**Comment by dougbu** [comment]
Nice find @chrisbecke❕ I reproduced this issue and am now debugging to find the cause&hellip;

**Comment by dougbu** [comment]
Fix is available in #35767. Note however that I should add a test for this scenario.

**Comment by dougbu** [comment]
- [ ] #36171
- [ ] #36355
- [ ] #36363
- [ ] #36362
